### PR TITLE
Feature/505 add fixed dataset for validation in experiment

### LINKED ISSUE #505: Allow for fixed validation dataset in Experiment
In GitLab by @sharkovsky on Apr 14, 2023, 15:56

As a data scientist, I want to be able to reuse the same validation dataset across rounds (instead of randomly sampling a new validation dataset each round).

### PR DESCRIPTION
**Pull request by ybouilla**
**PR description**

Closes #505 

Adds new feature for keeping same testing dataset over several `Rounds`

This feature is based on saving training and testing indices at the end of each `Round`, and load them at the beginning of each new `Round`

**Developer Certificate Of Origin (DCO)**

By opening this merge request, you agree to the
[Developer Certificate of Origin (DCO)](https://github.com/fedbiomed/fedbiomed/blob/master/CONTRIBUTING.md#fed-biomed-developer-certificate-of-origin-dco)

This DCO essentially means that:

- you offer the changes under the same license agreement as the project, and
- you have the right to do that,
- you did not steal somebody else’s work.

**License**

Project code files should begin with these comment lines to help trace their origin:
```
# This file is originally part of Fed-BioMed
# SPDX-License-Identifier: Apache-2.0
```

Code files can be reused from another project with a compatible non-contaminating license.
They shall retain the original license and copyright mentions.
The `CREDIT.md` file and `credit/` directory shall be completed and updated accordingly.

**Guidelines for PR review**

General:

* give a glance to [DoD](https://fedbiomed.org/latest/developer/definition-of-done/)
* check [coding rules and coding style](https://fedbiomed.org/latest/developer/usage_and_tools/#coding-style)
* check docstrings (eg run `tests/docstrings/check_docstrings`)

Specific to some cases:

* update all conda envs consistently (`development` and `vpn`, Linux and MacOS)
* if modified researcher (eg new attributes in classes) check if breakpoint needs update (`breakpoint`/`load_breakpoint` in `Experiment()`, `save_state_breakpoint`/`load_state_breakpoint` in aggregators, strategies, secagg, etc.)
* if modified a component with versioning (config files, breakpoint, messaging protocol) then update the version following the rules in `common/utils/_versions.py`

### CODE REVIEW

File: fedbiomed/common/data/_sklearn_data_manager.py [hunk 1]
```diff
@@ -317,23 +359,130 @@ def split(self, test_ratio: float, test_batch_size: int) -> Tuple[NPDataLoader,
...
+            test_ratio = None
+        rng: None | np.random.Generator = None
+        # if rand_state is not None:
+        #     if rand_state['bit_generator'] == 'PCG64':
+        #         rng = np.random.PCG64()
+        #         rng.state = rand_state
+        #         data_loader_state['random_seed'] = None
+        #     else:
+        #         # raise error
+        #         pass
```
Comment (srcansiz) [inline fedbiomed/common/data/_sklearn_data_manager.py:455]: is this code intended to be used ?

File: fedbiomed/common/data/_sklearn_data_manager.py [hunk 2]
```diff
@@ -317,23 +359,130 @@ def split(self, test_ratio: float, test_batch_size: int) -> Tuple[NPDataLoader,
...
+                  + f"Hence, dataset will be reshuffled. More details: {e}"
+            logger.error(msg)
+            raise IndexError(msg)
+    # @staticmethod
+    # def _bit_generator_to_randomstate(rng: np.random.Generator | None) -> None | np.random.RandomState:
+    #     return rng and np.random.RandomState(rng.bit_generator)
+
+    # @staticmethod
+    # def _randomstate_to_bit_generator(rs: np.random.RandomState | None) -> None | np.random.Generator:
+    #     return rs and rs._bit_generator
```
Comment (srcansiz) [inline fedbiomed/common/data/_sklearn_data_manager.py:484]: > is this code intended to be used ? 

same comment

File: fedbiomed/common/data/_torch_data_manager.py [hunk 3]
```diff
@@ -162,6 +222,58 @@ def _subset_loader(self, subset: Subset, **kwargs) -> Union[DataLoader, None]:
 
         return self._create_torch_data_loader(subset, **kwargs)
 
+    def rng(self, rng: Optional[int] = None, device: Optional[str | torch.device] = None) -> Tuple[int | None, torch.Generator | None]:
+        self._rng = rng and torch.Generator(device).manual_seed(rng)
+        return self._rng
+
+    def _load_indexes(self, training_index: List[int], testing_index: List[int]):
+        # TODO: put this method in main class
```
Comment (srcansiz) [inline fedbiomed/common/data/_torch_data_manager.py:230]: Hi @ybouilla,

Which one is the main class? If you mean that there could be a main class that both torch and sklearn classes can extend, I definitely agree that idea.

File: fedbiomed/common/data/_torch_data_manager.py [hunk 4]
```diff
@@ -162,6 +222,58 @@ def _subset_loader(self, subset: Subset, **kwargs) -> Union[DataLoader, None]:
...
+
+    @classmethod
+    def load_state(clf, dataset, **state) -> 'TorchDataManager':
+        # if 'rng' in state:
+        #     rng_save = state.pop('rng')
+        # else:
+        #     rng_save = None
+        # clf(dataset, **state)
+        # rng = torch.Generator()
+        # clf._rng = rng.set_state(rng_save)
```
Comment (srcansiz) [inline fedbiomed/common/data/_torch_data_manager.py:259]: I think this code is intended to be removed before merging?

File: fedbiomed/common/data/_sklearn_data_manager.py [hunk 5]
```diff
@@ -317,23 +359,130 @@ def split(self, test_ratio: float, test_batch_size: int) -> Tuple[NPDataLoader,
...
+        _loader_args = {}
+        # if self._rng:
+        #     _loader_args['random_seed'] = self._bit_generator_to_randomstate(self.rng()).get_state()
+        _loader_args['training_dataset'], _loader_args['testing_dataset'] = self._get_indexes()
+        _loader_args['test_ratio'] = self._test_ratio
+
+        return _loader_args
+
+    @classmethod
+    def load_state(
```
Comment (srcansiz) [inline fedbiomed/common/data/_sklearn_data_manager.py:429]: HI @ybouilla,

I didn't see where these methods are used? I was expecting that it could be used in round to load the previous state but I couldn't see it. Did I miss it? Same for `save_sate` method?

**Review by srcansiz — COMMENTED** [review]
HI @ybouilla,

Thank you for PR! Please find my first comments. I haven't run an example yet but I will and add more review comments if necessary. 

Thanks!

File: fedbiomed/common/data/_data_manager.py [hunk 6]
```diff
@@ -57,9 +57,10 @@ def extend_loader_args(self, extension: Dict):
         Args:
             extension: the mapping used to extend the loader arguments
         """
-        self._loader_arguments.update(
-            {key: value for key, value in extension.items() if key not in self._loader_arguments}
-        )
+        if extension:
+            self._loader_arguments.update(
+                {key: value for key, value in extension.items() if key not in self._loader_arguments}
+            )
```
[the remark this answers is not in the document]
Comment (ybouilla) [reply fedbiomed/common/data/_data_manager.py:63]: Yes `extension` can be `None`, making the above failing

File: fedbiomed/common/data/_sklearn_data_manager.py (hunk 5 above)
Comment (ybouilla) [reply fedbiomed/common/data/_sklearn_data_manager.py:429]: Oh you are right... I forgot to add them in the `Round` method. I will try to fix that

**Review by ybouilla — COMMENTED** [review]
Thanks for your review! I will update my PR as soon as possible!

File: fedbiomed/common/data/_sklearn_data_manager.py (hunk 5 above)
Comment (ybouilla) [reply fedbiomed/common/data/_sklearn_data_manager.py:401]: Fixed!

**Comment by srcansiz** [comment]
Hi @ybouilla,
Could you please update the documentation "researcher/model-testing-during-federated-training" with this new behavior of the testing functionality. Additionally,  I couldn't find a basic tutorial where testing functionality is shown. What do you thing adding it to PyTorch MNIST example? It can be nice to add it there since it a basic and common functionality that everyone uses in every training.

**Comment by ybouilla** [comment]
> Could you please update the documentation "researcher/model-testing-during-federated-training" with this new behavior of the testing functionality. 
True I have completely  forgotten updating the docs! 
> Additionally, I couldn't find a basic tutorial where testing functionality is shown. What do you thing adding it to PyTorch MNIST example? It can be nice to add it there since it a basic and common functionality that everyone uses in every training.
I don't know, we didnot have any tutorial focusing on validation. I know it is used in the transfer learning tutorial though

File: fedbiomed/common/data/_torch_data_manager.py (hunk 4 above)
Comment (ybouilla) [reply fedbiomed/common/data/_torch_data_manager.py:233]: I was thinking it could belong to a common method shared with breakpoint facility: indeed, we dont handle correct reset of random seed on `Researcher` as well when restarting from a breakpoint

**Comment by srcansiz** [comment]
> > Could you please update the documentation "researcher/model-testing-during-federated-training" with this new behavior of the testing functionality.
> 
> True I have completely forgotten updating the docs!
> 
> > Additionally, I couldn't find a basic tutorial where testing functionality is shown. What do you thing adding it to PyTorch MNIST example? It can be nice to add it there since it a basic and common functionality that everyone uses in every training.
> 
> I don't know, we didnot have any tutorial focusing on validation. I know it is used in the transfer learning tutorial though

@sharkovsky, @mvesin, @lchambon  what do you think about including testing/validation instructions in basic PyTorch tutorial in docs?

**Comment by ybouilla** [comment]
Hey @srcansiz ,  I ve seen you removed the comment for setting reusable random seed. I am ok with that, but don't you want to save the code in a issue (in order to keep the work in someplace). It is not something straightforward

**Comment by srcansiz** [comment]
> Hey @srcansiz , I ve seen you removed the comment for setting reusable random seed. I am ok with that, but don't you want to save the code in a issue (in order to keep the work in someplace). It is not something straightforward

Hi @ybouilla, 

Yes sure let's create an issue for that but could you please re-explain when and where it will used in the future so I can mention in the issue.

**Comment by srcansiz** [comment]
Hi @ybouilla,
After the documentation is added the PR is good to merge for me! Thank you for the great work!
